Desktop app

spartDev opened the pull request:
<!-- FILL OUT THE BELOW FORM OR YOUR PR WILL BE AUTOMATICALLY CLOSED -->
**Status**
- [ ] WIP
- [x] Ready for review
- [x] Needs testing

**Release notes for users (delete if codebase-only change)**
- Create a desktop app vesrion of spectrum

closes: #257

spartDev committed:
create desktop application

make code smart

remove useless package

use the prod optimized build when not in dev mode

optimze electron dev

create packager script for mac

add electron app icons

disabled nodeIntegration

add packager script for linux

add packager script for windows

create electron app

mxstbr reviewed (commented):
This is a great start already!

The main big change I (with no Electron experience) can see here is that we really want the frontend to be bundled into the Electron app, rather than loaded as a remote URL, right?

spartDev commented:
Yes I think loaded as a remote URL it's a good idea, because the app is universal and we do not want to ship node server with the app imo

mxstbr commented:
> Yes I think loaded as a remote URL it's a good idea, because the app is universal and we do not want to ship node server with the app imo

We don't need to ship a Node server with the app if we bundle the frontend into Electron, since it'll just query api.spectrum.chat, right?

spartDev commented:
Yes my bad you totally right

spartDev commented:
We start having something !

For now I have created a packager script to ship only for mac
You can test it by running:
```js
yarn electron:package-mac
```
then go to `build-electron` folder and launch `Spectrum.app`

I'm facing an issue right now (open the dev tools):
<img width="515" alt="screen shot 2018-04-11 at 20 09 54" src="https://user-images.githubusercontent.com/531935/38635300-38e220f0-3dc5-11e8-915f-2f546e43c52e.png">

spartDev commented:
problem solved, we need to disable Electron's Node integration.
https://github.com/atom/electron/issues/1753

spartDev commented:
I rebase my PR for more more clarity.
For now we have: 

- A preload script to create a “bridge” between Electron and the remote spectrum web app.
- A badge count for notifications
- A splash screen


I still have to work on the autoupdater.

mxstbr commented:
Amazing—I'll test this locally!

mxstbr commented:
> I still have to work on the autoupdater.

This will require us to host some sort of server, right? I heard some folks are using GitHub releases for that purpose, is that an option for us?

mxstbr committed:
Add missing dependency

spartDev commented:
We can use GitHub Releases, Amazon S3, DigitalOcean Spaces or generic HTTP(s) server). I think GitHub Releases is a good option for us. Maybe @nikgraf who seems to have a very good xp in electron can help us on this ?

spartDev committed:
add app autoUpdate

spartDev committed:
remove unused require

spartDev committed:
rename to desktop  & create build et publish scripts

spartDev commented:
OK the real things begins !! 

I rename the folder from `electron` to `desktop` for more consistency.
I use [electron-builder](https://www.electron.build/) who is a complete solution to package, build & publish electron app.

I had to move previous scripts from `package.json` to `desktop/package.json` because `electron-builder` have a small conflict with `CRA`, both use a `build`folder.

The only thing I do not like is that the `package.json` file in `desktop` folder need to have a `name` and `version` set (mandatory by electron-builder). It's redundant but I did not find better ...
If anyone has a better idea !!

You can test now.

spartDev commented:
Oh the last thing to do is application code signing

spartDev committed:
dmg installer enhancement

spartDev committed:
fix dev environement

spartDev committed:
fix splash screen

spartDev committed:
add about dialog

spartDev committed:
rename dev:elctron script by dev:desktop

spartDev committed:
fix missing resources folder in release build

spartDev commented:
I have fix several bug on dev environment, fix the splash screen who did not appear and create an about dialog.

<img width="1402" alt="screen shot 2018-04-17 at 00 25 40" src="https://user-images.githubusercontent.com/531935/38838397-2c402940-41d6-11e8-9499-b647654ceb0a.png">

mxstbr commented:
This build process is super nice! I ran it locally and things work perfectly except for a couple bugs:

- [ ] After opening, it tried to redirect me to a non-existant channel
- [ ] I can't seem to enable web push notifications?

<img width="1067" alt="screen shot 2018-04-19 at 16 53 22" src="https://user-images.githubusercontent.com/7525670/38999520-4acf488c-43f2-11e8-89d4-da62598f8294.png">

This is the error in the console:

> "DOMException: Registration failed - push service not available"

mxstbr commented on desktop/package.json:
```diff
@@ -0,0 +1,73 @@
+    "wait-on": "^2.1.0"
+  },
+  "scripts": {
+    "dev": "cd ../ && concurrently \"cross-env BROWSER=none yarn dev:web\" \"wait-on http://localhost:3000 && electron ./desktop/src/main.js\"",
```
Let's make the dev script `electron ./desktop/src/main.js` and tell folks in the docs that they have to run the web client locally in a separate tab to develop the Electron app.

spartDev replied:
sound fine to me !

mxstbr commented on desktop/package.json:
```diff
@@ -0,0 +1,73 @@
+  },
+  "build": {
+    "productName": "Spectrum",
+    "appId": "com.spectrum",
```
This should probably be `chat.spectrum`?

spartDev replied:
Yes you're right, it's better

mxstbr commented on desktop/package.json:
```diff
@@ -0,0 +1,73 @@
+      "icon": "resources/icons/mac/icon.icns"
+    },
+    "dmg": {
+      "background": "resources/background.tiff",
```
I think I'd prefer a white background here with our purple logomark at the top, instead of the purple gradient with the white logomark at the top? The purple background really fights with the purple icon. /cc @uberbryn we'll probably need some assets for this.

<img width="652" alt="screen shot 2018-04-19 at 16 56 47" src="https://user-images.githubusercontent.com/7525670/38999695-acf79f8c-43f2-11e8-9cf6-9e8c2e67f58e.png">

spartDev replied:
Totally agree this was just a test, need to work with @uberbryn for that !

spartDev commented:
Thx for feedback. I'm on it !

superbryntendo commented:
@spartDev FYI I'm working on icons for this rn and I have a background texture and logo for the "drag to applications view". What size should those be?

superbryntendo commented:
(I've got the icon sizes, just need the bg image size really)

superbryntendo commented:
also @mxstbr, i thought com.xxxx is just the apple application naming scheme. Doesn't have to do with the URL afaik?

mxstbr commented:
> Doesn't have to do with the URL afaik?

I thought the whole point was that it was the inverse URL haha, but maybe I'm wrong!

spartDev commented:
it's the application used as [CFBundleIdentifier](https://developer.apple.com/library/content/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-102070) for MacOS and as [Application User Model ID](https://msdn.microsoft.com/en-us/library/windows/desktop/dd378459(v=vs.85).aspx) for Windows (NSIS target only, Squirrel.Windows not supported). I think the `com` is a convention

superbryntendo commented:
![image](https://user-images.githubusercontent.com/4926824/39007220-24ccad7e-43ba-11e8-9125-eb4981fad75f.png)

you're right! I'd always seen that in the Library folder but always thought it was just a mac nomenclature.

superbryntendo commented:
So, for the unique bundle identifier, let's just do `chat.spectrum.client` I guess.

mxstbr commented:
>  just need the bg image size really

I think that can be whatever you want? From the code it looks like the icons are positioned with coordinates

spartDev commented:
@uberbryn Icon size should be at least 512x512. we need 3 formats:

- icon.icns for mac
- icon.png fro linux
- icon.ico fro windows

About background for DMG: the resolution of this file determines the resolution of the installer window.
Default locations expected background size to be 540x380.
The common approach is a .tiff with normal size and @x2 size => https://developer.apple.com/library/content/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Optimizing/Optimizing.html#//apple_ref/doc/uid/TP40012302-CH7-SW10

superbryntendo commented:
I've already got all the icon sizes - that bg info is exactly what I needed. Thanks!

fullstackguyhere commented:
Hey, its cool that a desktop app is being developed. Do we have a plan for integrating all the features into desktop component....I am asking because I am eager to contribute!

brianlovin commented:
@fullstackguyhere desktop app is mostly just wrapping the website, so should match all the functionality by default!

superbryntendo commented:
@spartDev 

Icons (mac, win, png - all sizes):

[Dock Icons.zip](https://github.com/withspectrum/spectrum/files/1930253/Dock.Icons.zip)

Installer bg for mac (540x380 @2x):

[pkg-bg.zip](https://github.com/withspectrum/spectrum/files/1930255/pkg-bg.zip)

spartDev commented:
@uberbryn thx for assets, just a couple things:

- The icon is not transparent
![screen shot 2018-04-20 at 11 13 43](https://user-images.githubusercontent.com/531935/39042951-78b2dc06-448c-11e8-8e4c-7ff168adbd55.png)

- The app installer BG seems not to be good
![screen shot 2018-04-20 at 11 14 05](https://user-images.githubusercontent.com/531935/39043121-de4964ea-448c-11e8-9fb6-3f128282df63.png)

Can you create 2 TIFF files, one with a size of **540 x 380** (for non retina) and one other **@x2** (for retina) then run this TIFF Utility Command in your Terminal:
```bash
tiffutil -cathidpicheck background.png background@2x.png -out background.tiff
```
this command will produce a single TIFF file that contains the two input images.
[Link to apple doc](https://developer.apple.com/library/content/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Optimizing/Optimizing.html#//apple_ref/doc/uid/TP40012302-CH7-SW13)

About icon shape, this does not seem consistent with [apple guideline](https://developer.apple.com/macos/human-interface-guidelines/icons-and-images/app-icon/

Maybe this can help:
[Yosemite_Icon_Template.sketch.zip](https://github.com/withspectrum/spectrum/files/1931575/Yosemite_Icon_Template.sketch.zip)

spartDev committed:
update appId to be more consistent

spartDev committed:
simplify the development script

spartDev committed:
update documentation

spartDev committed:
fix desktop dev script

spartDev commented:
@mxstbr 

> I can't seem to enable web push notifications?

unfortunately `PushManager` is missing in Electron see [issue](https://github.com/electron/electron/issues/6697)

I found this [electron-push-receiver](https://github.com/MatthieuLemoine/electron-push-receiver) but is only for Firebase Cloud Messaging

I keep digging in it

superbryntendo commented:
@spartDev Hmmm. idk what happened with transparency there. that sucks...

I'm planning to add the traditional macos shadows, etc. but wanted to see it in situ first. I'm not making it a circle as that doesn't fit our brand and most electron apps don't follow that guideline anyway. Abstract's icon is a good example of what I'm going for.

I'll update the bg sizes

superbryntendo commented:
I slapped it onto another app to test it out. After a whole bunch of experimenting, I'm pretty happy with this now:
![image](https://user-images.githubusercontent.com/4926824/39087906-2e33c696-455d-11e8-8b9a-4a7c818fb2fb.png)
[Dock Icons.zip](https://github.com/withspectrum/spectrum/files/1934893/Dock.Icons.zip)

superbryntendo commented:
Updated bg:
[background.tiff.zip](https://github.com/withspectrum/spectrum/files/1934898/background.tiff.zip)

spartDev committed:
add new icons

spartDev committed:
update background dmg installer

spartDev commented:
@uberbryn works perfectly !
![screen shot 2018-04-22 at 12 18 53](https://user-images.githubusercontent.com/531935/39093931-a4f4092e-4627-11e8-9ce6-3d80a9d8549a.png)

brianlovin commented:
😍😍😍

mxstbr commented:
That looks awesome! So now, the last bit is figuring out how to handle push notifications, right?

spartDev commented:
@mxstbr Yes that's it.
I dug in this weekend, unfortunately Push Notification is missing in Electron.
I'm not a push notification expert, currently you use your own server for push notification not **Firebase** or **GCM** ? 
A work around like Slack we can use **websocket**, or simply send notifications using regular **Notifications API**.

mxstbr commented:
> A work around like Slack we can use websocket, or simply send notifications using regular Notifications API.

We already listen to new notifications via websockets, can we just send push notification once one comes in?

spartDev commented:
I'm on it !

spartDev commented:
@mxstbr After some debug in app I saw that you were going through Firebase Cloud messaging.
The good news is that there is a module to bring Web Push support to Electron: https://github.com/MatthieuLemoine/electron-push-receiver
But need some configuration like SERVER_KEY Authorization ...

mxstbr commented:
> I saw that you were going through Firebase Cloud messaging

Uhh that's not true, we have our own notifications infrastructure?

spartDev commented:
Maybe I completely wrong but when I logged the promise of `WebPushManager.subscribe()` in notification settings I saw that:
![screen shot 2018-04-25 at 20 10 20](https://user-images.githubusercontent.com/531935/39264467-3d5a81da-48c5-11e8-9364-1553b4a57037.png)

mxstbr commented:
Yeah, that's just because Google uses the same servers to send Web Push Notifications, but we send them from our own servers, not via the firebase library

https://github.com/withspectrum/spectrum/blob/fac69fc8475ce2c22cc0f33a16a5b89a2ae4ab94/shared/send-web-push-notification.js#L39-L42

If you try and register for web push notifications in Firefox that'll log an entirely different server, hosted by Mozilla!

spartDev commented:
Oh !! thanks, it's more clear to me.

terry-fei commented:
Is there any progress？

spartDev commented:
I am on vacation now I finish on my return but any contribution us welcome
😉
Le ven. 4 mai 2018 à 04:47, afei <notifications@github.com> a écrit :

> Is there any progress？
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/withspectrum/spectrum/pull/2825#issuecomment-386493568>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAgd33kGUJx7_TWVNrrPwQsGpt1aUyAvks5tu8FKgaJpZM4TP9fo>
> .
>
-- 
Thomas Roux

superbryntendo committed:
Merge branch 'alpha' into electron-wrapper

mxstbr committed:
Merge branch 'alpha' into electron-wrapper

mxstbr commented:
I'm tackling push notifications, let's get this ready to be beta tested! :v:

mxstbr committed:
Fix variable name

mxstbr committed:
Implement push notifications in desktop app

My initial idea was to use our existing NavBar component, which is
already listening to new notifications, and to show push notifications
from there. That was way more complex than necessary honestly, so I just
manually listen to the subscriptions in `src/index.js` and it's working
super well.

This now shows push notifications for all desktop users! :tada:

mxstbr committed:
Merge branch 'alpha' into electron-wrapper

mxstbr commented:
The latest commits implement push notifications in the desktop app! :tada: Let's test the crap outta these, here's are Mac builds with the latest code, download them and let us know of any bugs:

**TODO BEFORE MERGING**

- [ ] Fix Linux and Windows builds (they currently fail with some icon error)
- [ ] Test the shit outta it

mxstbr commented:
The latest commits implement push notifications in the desktop app! :tada: 

**TODO BEFORE MERGING**

- [x] Don't send push notifications if the user's looking at the app
- [x] Fix Linux and Windows builds (they currently fail with some icon error)
- [x] Figure out all configuration options necessary for the best experience
- [x] Set min-width to 770px to avoid people going into the mobile view
- [x] Set min-height to 600px (or maybe even smaller) for smaller screens
- [ ] Fix `isElectron` check to not trigger in Cypress
- ~~Implement `hasReply` into push notifications~~ Not possible as we dispatch push notifications from the renderer process, not the main process.

mxstbr committed:
Add build:desktop

mxstbr committed:
Fix unit tests for notification-to-text util

mxstbr committed:
Dont send push notifications if the desktop app is focused

mxstbr committed:
Fix linux build

mxstbr committed:
Fix windows build

mxstbr committed:
Build desktop apps on CircleCI

mxstbr committed:
Fix yaml

mxstbr committed:
Install desktop deps on circle

mxstbr committed:
Adapt electron builder options

mxstbr committed:
Remove accidentally committed patch

mxstbr committed:
Cleanup

mxstbr committed:
Fix cannot read property error

mxstbr committed:
Fix width and height values

mxstbr committed:
Fix ci error

mxstbr committed:
Make sure were not running in Cypress

mxstbr committed:
Move desktop push sending method out of src/index.js

mxstbr commented:
I think this is pretty much ready! @brianlovin mind giving this a last code review, then doing a prod push with the frontend changes of this branch?

As soon as those are live, anybody who has a build of the desktop app should start seeing push notifications (if the app is open) :tada:

brianlovin commented:
Prod cut incoming!

brianlovin commented:
Thanks so much @spartDev - this has been awesome to use so far
